zerver/actions/streams.py

```diff
@@ -133,6 +133,67 @@ def do_deactivate_stream(
     )
 
 
+def bulk_delete_cache_keys(message_ids_to_clear: List[int]) -> None:
+    while len(message_ids_to_clear) > 0:
+        batch = message_ids_to_clear[0:5000]
+
+        keys_to_delete = [to_dict_cache_key_id(message_id) for message_id in batch]
+        cache_delete_many(keys_to_delete)
+
+        message_ids_to_clear = message_ids_to_clear[5000:]
+
+
+def merge_streams(realm: Realm, stream_to_keep: Stream, stream_to_destroy: Stream) -> List[int]:
+    recipient_to_destroy = stream_to_destroy.recipient
+    recipient_to_keep = stream_to_keep.recipient
+    if recipient_to_destroy.id == recipient_to_keep.id:
+        return [0, 0, 0]
+
+    # gather info before taking action
+    existing_subs = Subscription.objects.filter(recipient=recipient_to_keep)
+    users_already_subscribed = {sub.user_profile_id: sub.active for sub in existing_subs}
+    message_ids_to_clear = list(
+        Message.objects.filter(recipient=recipient_to_destroy).values_list("id", flat=True)
+    )
+    subs_to_deactivate = Subscription.objects.filter(recipient=recipient_to_destroy, active=True)
+    users_to_activate = [
+        sub.user_profile
+        for sub in subs_to_deactivate
+        if not users_already_subscribed.get(sub.user_profile_id, False)
+    ]
+
+    # follow a safe order of operations, see
+    # https://github.com/zulip/zulip/pull/22262#pullrequestreview-1017884964
+
+    # step 1. Subscribe people to the new stream
+    # note: this simple algorithm doesn't preserve any stream settings/colors/etc.
+    if len(users_to_activate) > 0:
+        bulk_add_subscriptions(realm, [stream_to_keep], users_to_activate, acting_user=None)
+
+    # step 2. Move the messages
+    # also, delete the old copies from caches.
+    count = Message.objects.filter(recipient=recipient_to_destroy).update(
+        recipient=recipient_to_keep
+    )
+    bulk_delete_cache_keys(message_ids_to_clear)
+
+    # step 3. Unsubscribe everyone from the old stream
+    if len(subs_to_deactivate) > 0:
+        bulk_remove_subscriptions(
+            realm,
+            [sub.user_profile for sub in subs_to_deactivate],
+            [stream_to_destroy],
+            acting_user=None,
+        )
+
+    # step 4. Deactivate the stream
+    # note: this might fail in the case that the stream is used by the system, e.g. for
+    # signups, notifications, etc. In this case, in the admin settings either disable these
+    # features or change the stream to something else.
```

Comment:
This note is useful as a comment, but I might go further than "might fail" and say "might raise a JsonableError" so that it's explicit the _way_ in which it will fail.

But this also confuses me, since below you have a test which shows that merging the notifications stream into the Denmark stream (destroying the notifications stream) is successful, with no exception.  Which is it -- does that raise an exception, or does that unset the notifications stream?

Reply:
DONE - removed the confusing comment. Sorry, this is a pre-commit before the report-message PR #21821, where it does throw an exception for attempting to archive the reporting stream ([link](https://github.com/zulip/zulip/pull/21821/files#diff-cd6028ed9056ad3e83e6a37f21dfaf1311e45266dc3dc38b710f5f01ab5ec65fR450))

Once report-message is committed, I could see taking on a little refactoring to elevate notification-type-streams to a known type of object with certain properties (e.g. cannot be archived)...